### PR TITLE
Add engine release instructions.

### PR DESCRIPTION
**Pull request by sipsma**
Signed-off-by: Erik Sipsma <erik@sipsma.dev>

**There are a few parts of the instructions I'd like to automate ASAP (before official release ideally), will comment on them. Just writing out the instructions as is for now.**

### CODE REVIEW

File: RELEASING.md [hunk 1]
```diff
@@ -1,34 +1,74 @@
...
+
 ```console
 git pull
 git status  # make sure everything is clean
 ```
 
-### Release Engine
+### Bump version tags
+
+- Modify `EngineImageRef` in `internal/mage/engine.go` to use a tag set to the engine version (e.g. `v0.3.0`).
```
Comment (sipsma) [inline RELEASING.md:21]: I think this is automatable. Can have the code look at the current git branch and use that to set the tag. Will need to add some sort of confirmation dialog too. I think that slight extra bit of complication would be worth it to get this more automated, but let me know if you disagree

File: RELEASING.md [hunk 2]
```diff
@@ -1,34 +1,74 @@
...
 
+1. Ensure all tests in CI pass.
+1. Additionally, manually test provisioning by running the following commands on both Linux and MacOS hosts directly
+
+```console
+export SHORT_DIGEST=a10730b1d305644b # replace the value with the first 16 chars of the digest of the engine image used above
+docker rm -fv dagger-engine-$SHORT_DIGEST # ensure there's no existing image so we use the full provisioning code paths
+rm -rf ~/.cache/dagger/* # ensure there's no existing local engine-session binaries for same reason as above
+cd ../.. # go to the root of the dagger repo
+go test -v -count=1 $(pwd)/core/integration # run the engine tests on your host, which exercises the provisioning code paths of the Go SDK
```
Comment (sipsma) [inline RELEASING.md:69]: This should be fairly easy to automate as a mage step (the reason it isn't already is because we run tests dagger-in-dagger, which is great but skips the provisioning code paths)

Comment (sipsma) [reply RELEASING.md:69]: Though automating it on MacOS is not going to happen before the next release, so that will still be manual

File: RELEASING.md [hunk 3]
```diff
@@ -39,12 +79,51 @@ git push origin sdk/go/${SDK_VERSION}
...
+```
+
+1. Ensure all tests in CI pass.
+1. Additionally, manually test provisioning by running the following commands on both Linux and MacOS hosts directly
+
+```console
+export SHORT_DIGEST=a10730b1d305644b # replace the value with the first 16 chars of the digest of the engine image used above
+docker rm -fv dagger-engine-$SHORT_DIGEST # ensure there's no existing image so we use the full provisioning code paths
+rm -rf ~/.cache/dagger/* # ensure there's no existing local engine-session binaries for same reason as above
+poetry run poe test # to be run the `sdk/python` directory in our dagger repo
```
Comment (sipsma) [inline RELEASING.md:122]: Would like to automate this part, but not sure how straightforward it is when we don't rely on dagger-in-dagger. Will give it a shot, but like the Go SDK, will be at least partially manual on MacOS for now.

**Comment by sipsma** [comment]
Also, I want to do a double check of the instructions by running through them all manually (not actually pushing any tags) before merging this. But let me know what feedback there is in the meantime

File: RELEASING.md (hunk 1 above)
Comment (gerhard) [reply RELEASING.md:21]: Yes, that is exactly what @aluzzardi mentioned earlier today - `mage sdk:python:engine:bump` - so we are all on the same page.

File: RELEASING.md [hunk 4]
```diff
@@ -1,34 +1,74 @@
 # Releasing
 
-## Go SDK
+> **Warning**
+> Currently, if any SDK bumps an engine dependency, we should bump the engine dependency of all other SDKs concurrently. This helps ensure that users running multiple SDKs from a single machine all use the same engine. This is important as currently our automatic provisioning code enforces the existence of a single engine running at a time.
+
```
Comment (gerhard) [inline RELEASING.md:5]: **This is super important**. I thought that multiple engines with different versions are supported. Great to know this!

cc @aluzzardi @marcosnils

File: RELEASING.md [hunk 5]
```diff
@@ -1,34 +1,74 @@
...
-git commit go.mod go.sum -m 'sdk: go: bump engine dependency'
+```
+
+1. Obtain the full image ref, including SHA, from our GHCR repo for the version of the engine the SDK should depend on.
+
+- Our GHCR repo can be found [here](https://github.com/orgs/dagger/packages/container/package/engine).
+- Locate the tag for the engine the SDK should depend on and note the digest of it.
+  - The digest will be prefixed with `sha256:` and look like `a10730b1d305644b87c21630d27d9e15d424014ee7ac9bd44f4504fcfe92a8ca`
+
+1. Update the value of `engineImageRef` in `client.go` to be `ghcr.io/dagger/engine:<version tag>@sha256:<digest>` ([line currently here](https://github.com/dagger/dagger/blob/349f8a1661f1a8422a8bef919a345bd6dd3f3d85/sdk/go/client.go#L19))
```
Comment (gerhard) [inline RELEASING.md:54]: This is my favourite image ref format too: `<version>@sha256:<digest>`

I propose that we default to this whenever possible.

File: RELEASING.md (hunk 4 above)
Comment (sipsma) [reply RELEASING.md:5]: Yes, more context in this comment and the one right above it: https://github.com/dagger/dagger/pull/3647#issuecomment-1302721770

tl;dr it's a simple solution for now, we can find ways of improving it in the near future (at the cost of some extra complexity)

File: internal/mage/engine.go [hunk 6]
```diff
@@ -66,13 +66,15 @@ func (t Engine) Publish(ctx context.Context) error {
...
-	imageRef, err := c.Container().Publish(ctx, engineImageRef, dagger.ContainerPublishOpts{
+	imageRef, err := c.Container().Publish(ctx, EngineImageRef, dagger.ContainerPublishOpts{
 		PlatformVariants: util.DevEngineContainer(c, arches, oses),
 	})
 	if err != nil {
 		return err
 	}
-	fmt.Println("Image published:", imageRef)
+
+	time.Sleep(3 * time.Second) // allow buildkit logs to flush, to minimize potential confusion
```
Comment (gerhard) [inline internal/mage/engine.go:76]: Is it possible to check when logs are flushed? If yes, I would check periodically and fail on timeout. Sleeps like this one get worse the longer they are kept around.

Comment (sipsma) [reply internal/mage/engine.go:76]: I'll double check quickly, but I think the answer is no, being able to check that will require a lot more changes to our logging related code (which should happen, just will take time0

File: internal/mage/engine.go [hunk 7]
```diff
@@ -6,12 +6,17 @@ import (
...
 
 	"dagger.io/dagger"
 	"github.com/dagger/dagger/internal/mage/util"
 	"github.com/magefile/mage/mg" // mg contains helpful utility functions, like Deps
 )
 
+const (
+	EngineImageRef = "ghcr.io/dagger/engine:v0.3.3"
+)
+
```
Comment (gerhard) [inline internal/mage/engine.go:19]: Not sure if this is intentional. If it is, can we do the same for the Python SDK in the same context?

https://github.com/dagger/dagger/blob/f47ac949b0acff657cf37c819c8ee4719d6f34fb/sdk/python/src/dagger/connectors/base.py#L19-L22

Comment (sipsma) [reply internal/mage/engine.go:19]: Ah okay, so *this* line in `internal/mage/engine.go` is where we have the code for publishing the image, so here we can't hardcode the SHA.

For the Go SDK we actually do hardcode the SHA same as we do in python (in `sdk/go/client.go`).

**Review by gerhard — APPROVED** [review]
Optimistic approval so that we can get a subset of this in (if possible).

How does this all look to you @aluzzardi?

**Comment by sipsma** [comment]
> Optimistic approval so that we can get a subset of this in (if possible).

Thanks @gerhard, I just pushed an extra commit with the first batch of automation (`engine:publish` runs in GHA on new engine tags now). If there's extra concerns introduced by this I'm happy to remove that commit from this PR, merge this as is and send out a separate one for the new automation, just figured it would be easier to consolidate while waiting for more eyes on this PR.

File: .github/workflows/publish-engine.yml [hunk 8]
```diff
@@ -0,0 +1,28 @@
+name: publish-engine
```
Comment (sipsma) [inline .github/workflows/publish-engine.yml:1]: All the config in here is just my best guess at what should work; I'm not actually sure how to test it before we actually do a new engine tag. Let me know if there's a good way of going about that.

Primary risks:
1. The `tags` matching patterns are wrong (just guessing based off reading the gha docs)
2. `secrets.RELEASE_DAGGER_CI_TOKEN` doesn't have the permissions needed to push to ghcr (https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry#authenticating-to-the-container-registry)

**Comment by sipsma** [comment]
Just FYI, pushing a temporary commit that enables the workflow on pull requests so that we can get some verification of the workflow now. Will remove before merge

**Review by aluzzardi — COMMENTED** [review]
(for later)

How about we move engine refs to a generated file per SDK, for example:

```
sdk/go/engine_ref.gen.go
sdk/python/dagger/engine_ref.gen.py
```

Then we can bump them programmatically, every time we release an engine:

```
mage sdk:all:bump <engine version>
[open PR]
```

File: .github/workflows/publish-engine.yml [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+on:
+  push:
+    branches:
+      - main
+    tags:
+      - v*
+      - "!sdk/go/**"
```
Comment (aluzzardi) [inline .github/workflows/publish-engine.yml:8]: This should not be needed -- I hope?

Comment (sipsma) [reply .github/workflows/publish-engine.yml:8]: My thinking was that we don't want to rebuild a new engine image if all we did was tag a new sdk release. I.e. don't want to build an engine image like `ghcr.io/dagger/engine:sdk/go/v0.4.1` (I don't think that's even a valid image tag?).

Instead this workflow should only run when we do a new engine tag or on main.

Right? Easily could be missing something...

Comment (sipsma) [reply .github/workflows/publish-engine.yml:8]: Oh also, just to clarify, since we hardcode the image SHAs in the SDKs, the `tag` is not functionally important to the provisioning.

The only reason to tag with the engine version is that it makes it easier for us to track which engine image is associated with which version of the engine. Just reduces risk of human error when updating SDKs.

**Comment by sipsma** [comment]
Did a test run of the github workflow by pushing to the dagger repo with a temporary tag, everything worked as expected there, so probably as good of verification we'll get before merging. 

After merging this I'll make follow up issues for:
* More automated provisioning testing for each SDK as a release blocker (including MacOS)
* Automating SDK engine dep bumps as suggested by @aluzzardi here: https://github.com/dagger/dagger/pull/3749#pullrequestreview-1174722671

File: .github/workflows/publish-engine.yml (hunk 9 above)
Comment (aluzzardi) [reply .github/workflows/publish-engine.yml:8]: > Instead this workflow should only run when we do a new engine tag or on main.

I **think** that is captured by `v*` -- `sdk/go/vXXX` shouldn't match the `v*` pattern

File: .github/workflows/publish-engine.yml [hunk 10]
```diff
@@ -0,0 +1,28 @@
+name: publish-engine
+on:
+  push:
+    branches:
+      - main
```
Comment (aluzzardi) [inline .github/workflows/publish-engine.yml:5]: This will trigger the workflow on every merge to main.

Do we want that?

File: .github/workflows/publish-engine.yml (hunk 9 above)
Comment (sipsma) [reply .github/workflows/publish-engine.yml:8]: omg duh..... okay, that makes perfect sense. Yeah I also hope the rest aren't needed 😂 I will remove them.

File: .github/workflows/publish-engine.yml (hunk 10 above)
Comment (sipsma) [reply .github/workflows/publish-engine.yml:5]: It seemed like we may as well since it will allow us to do local testing of SDKs against pushed-but-not-released engine builds. It's not strictly necessary though, just seemed nice-to-have.

I guess the downside is we will be pushing a lot more data to our ghcr repo. If that's a significant downside or there's other ones I'll get rid of this, let me know.

Comment (aluzzardi) [reply .github/workflows/publish-engine.yml:5]: 👍

**Review by aluzzardi — APPROVED** [review]

**Comment by gerhard** [comment]
It was great to see this come together so quickly - amazing! 🏎 🏎
